readme.md

```diff
@@ -47,126 +62,497 @@ This package exports [ESM](https://gist.github.com/sindresorhus/a39789f98801d908
+  },
+  customExtractImplementation?: (params: z.infer<Schema>) => z.infer<Schema> | Promise<z.infer<Schema>>
+): ExtractFunction<Schema>
 ```
 
-[source](./examples/chatbot)
+#### createAIRunner
+
+Creates a function to run a chat model in a loop, handling parsing, running, and inserting responses for function & tool call messages.
+
+```typescript
+createAIRunner<Content = string>(args: {
+  chatModel: Model.Chat.Model;
+  functions?: AIFunction[];
+  shouldBreakLoop?: (msg: Msg) => boolean;
+  maxIterations?: number;
+  functionCallConcurrency?: number;
+  validateContent?: (content: string | null) => Content | Promise<Content>;
+  mode?: Runner.Mode;
+  systemMessage?: string;
+  onRetriableError?: (error: Error) => void;
+}): Runner<Content>
+```
+
+### Utilities
+
+#### MsgUtil
+
+Utility class for creating and checking message types.
+
+- `MsgUtil.system(content: string, opts?): Msg.System`
+- `MsgUtil.user(content: string, opts?): Msg.User`
+- `MsgUtil.assistant(content: string, opts?): Msg.Assistant`
+- `MsgUtil.funcCall(function_call: { name: string; arguments: string }, opts?): Msg.FuncCall`
+- `MsgUtil.funcResult(content: Jsonifiable, name: string): Msg.FuncResult`
+- `MsgUtil.toolCall(tool_calls: Msg.Call.Tool[], opts?): Msg.ToolCall`
+- `MsgUtil.toolResult(content: Jsonifiable, tool_call_id: string, opts?): Msg.ToolResult`
+
+#### Tokenizer
+
+Utility for encoding, decoding, and counting tokens for various models.
+
+- `createTokenizer(model: string): Tokenizer`
+
+#### Cache
+
+Utilities for caching model responses.
+
+- `type CacheStorage<KeyType, ValueType>`
+- `type CacheKey<Params extends Record<string, any>, KeyType = string>`
+- `defaultCacheKey<Params extends Record<string, any>>(params: Params): string`
+
+## Advanced Topics
+
+### OpenAI Client (openai-fetch)
+
+Dexter uses the `openai-fetch` library to interact with the OpenAI API. This client is lightweight, well-typed, and provides a simple interface for making API calls. Here's how it's used in Dexter:
+
+1. **Default Client**: By default, Dexter creates an instance of `OpenAIClient` from `openai-fetch` when initializing models.
+
+2. **Custom Client**: You can provide your own instance of `OpenAIClient` when creating a model:
+
+   ```typescript
+   import { OpenAIClient } from 'openai-fetch';
+   import { ChatModel } from '@dexaai/dexter';
+
+   const client = new OpenAIClient({ apiKey: 'your-api-key' });
+   const chatModel = new ChatModel({ client });
+   ```
+
+3. **Client Caching**: Dexter implements caching for `OpenAIClient` instances to improve performance when creating multiple models with the same configuration.
+
+4. **Streaming Support**: The `openai-fetch` client supports streaming responses, which Dexter utilizes for real-time output in chat models.
+
+5. **Structured Output**: Dexter supports OpenAI's structured output feature through the `createExtractFunction`, which uses the `response_format` parameter with a JSON schema derived from a Zod schema.
```

Comment:
6. Dexter supports **any** LLM compatible with the OpenAI chat completions API.

See https://gptlint.dev/guide/llm-providers for my docs on how people can use anthropic claude via openrouter or local models via `ollama` / `vllm` which we should at least touch on.

Reply:
This isn't something we're using and don't want to deal with the issue from. The intelligent user will figure it out if motivated :)